# Completing exercises from Assignment.md

In [2]:
""" 1. Create each of the following arrays and print its `shape`, `ndim`, and `dtype`:
   - A 1D array of integers from 10 to 19  
   - A 2D array of shape `(4, 3)` filled with 1.5  
   - A 3D array of zeros with shape `(2, 2, 3)` 
"""

import numpy as np

# 1D array of integers from 10 to 19
array_1d = np.arange(10, 20)
array_1d

array([10, 11, 12, 13, 14, 15, 16, 17, 18, 19])

In [3]:
# 2D array of shape (4, 3) filled with 1.5
array_2d = np.full((4, 3), 1.5)
array_2d

array([[1.5, 1.5, 1.5],
       [1.5, 1.5, 1.5],
       [1.5, 1.5, 1.5],
       [1.5, 1.5, 1.5]])

In [4]:
# 3D array of zeros with shape (2, 2, 3)
array_3d = np.zeros((2, 2, 3))
array_3d

array([[[0., 0., 0.],
        [0., 0., 0.]],

       [[0., 0., 0.],
        [0., 0., 0.]]])

In [64]:
# Additional from suggested solution: to print out the shapes, dim and dtype of all created arrays
for a in [array_1d, array_2d, array_3d]:
    print(f"Shape: {a.shape}, ndim: {a.ndim}, dtype: {a.dtype}")

Shape: (10,), ndim: 1, dtype: int64
Shape: (4, 3), ndim: 2, dtype: float64
Shape: (2, 2, 3), ndim: 3, dtype: float64


In [6]:
# 2. Convert the `(4, 3)` float array to integers using `.astype(int)`.  
array_2d_int = array_2d.astype(int)
array_2d_int.dtype

dtype('int64')

3. Write down (in a markdown cell) when you would prefer `float` vs `int` in real analyst work (e.g., prices vs counts).

For counts, it is usually discrete therefore using int is more optimal. Whereas for prices, it may be produce better accuracy to have the data in float, else the data may underestimate actual figures, since int type removes the decimal points (round down)

In [7]:
# 2. Sales table drill: Indexing and slicing

# Scenario:** You have quarterly sales numbers for 4 regions (rows) over 5 quarters (columns).

sales = np.array([
    [120, 135, 110, 145, 130],  # Region A
    [ 90,  85,  95, 100, 110],  # Region B
    [200, 215, 190, 225, 245],  # Region C
    [155, 140, 135, 160, 170]   # Region D
])
regions = np.array(["A", "B", "C", "D"])
quarters = np.array(["Q1", "Q2", "Q3", "Q4", "Q5"])


In [8]:
# 1. Select all quarters for **Region B** as a 1D array. 
sales_B = sales[1]          # Index 1 is the 2nd row, which is where Region B data is. This will return a 1D array of all quarters for Region B.
sales_B

# sales_B = np.array(sales[1, :])     # Index 1 is the 2nd row, which is where Region B data is. ":" selects values from all columns
# need not specify ":" as it is implied when we select a single row, but it can be included for clarity.

array([ 90,  85,  95, 100, 110])

In [9]:
# 2. Select **Q2 to Q4** (inclusive) for all regions as a 2D subarray. 
sales_all_Q2to4 = sales[:, 1:4]  # ":" selects all rows, "1:4" selects columns from index 1 to 3 (Q2 to Q4)
sales_all_Q2to4

array([[135, 110, 145],
       [ 85,  95, 100],
       [215, 190, 225],
       [140, 135, 160]])

In [16]:
# My attempt before copilot kicks in
sales_q5_ad = sales[(0, -1), -1]
sales_q5_ad

# The above () should not be used. It is tuple and not a list, and just work because of the example.
# To use []

array([130, 170])

In [18]:
# 3. Select **Q5** sales for **Regions A and D only** (use slicing or fancy indexing).  3. Select **Q5** sales for **Regions A and D only** (use slicing or fancy indexing).  
# Slicing example
sales_q5_AD_slice = sales[[0, 3], -1]  # "[0, 3]" selects rows for Regions A and D, "-1" selects the last column (Q5)
sales_q5_AD_slice
# So what I tried is slicing, but using "-1" instead of "3"

array([130, 170])

In [19]:
# Fancy indexing example
sales_q5_AD_fancy = sales[np.ix_([0, 3], [-1])]  # np.ix_ creates an index for the specified rows and columns
sales_q5_AD_fancy

array([[130],
       [170]])

Slicing involves ":", so co-pilot didn't suggest correctly. However in this example, we can force fulfilling slicing by using sales[::3, -1] since there are 4 rows, we get all, but skip two rows in between.

In [ ]:
# 4. Compute the **shape** of each result and add a brief comment: “Is this 1D or 2D, and why?”

# For Qn1, the shape of sales_B is (5,) as we only extracted a row of 5 results. Per question, it is 1D.
# For Qn2, since we get all regions (4) and 3 results from Q2 to 4 each, the shape is (4,3). Per question, it is 2D.
# For Qn3, we specifically extracted 2 regions, with 1x Q5 sales data each, the shape is (2,) if we use purely sales[[0, 3], -1], but if we use np.ix_ it will be (2, 1) as it retains the 2D structure of the original array. Per question, it is 1D and 2D respectively.

In [65]:
# Additional from suggested solution: to print out the shapes of all results
print(f"Shape of sales_B: {sales_B.shape} (1D)")
print(f"Shape of sales_all_Q2to4: {sales_all_Q2to4.shape} (2D)")
print(f"Shape of sales_q5_AD_slice: {sales_q5_AD_slice.shape} (1D)")

Shape of sales_B: (5,) (1D)
Shape of sales_all_Q2to4: (4, 3) (2D)
Shape of sales_q5_AD_slice: (2,) (1D)


## 3\. Boolean masks: Work with targets

In [20]:
# **Scenario:** You have campaign response data.

names = np.array(["Ana", "Ben", "Chen", "Dana", "Eli", "Fatima", "George", "Hui"])
spend = np.array([200, 150, 300, 120, 180, 220, 160, 310])      # marketing spend
revenue = np.array([400, 180, 500, 100, 220, 260, 150, 600])   # revenue

In [21]:
# 1. Compute the **ROI** for each person: `roi = revenue / spend`.  
roi = revenue / spend
roi

array([2.        , 1.2       , 1.66666667, 0.83333333, 1.22222222,
       1.18181818, 0.9375    , 1.93548387])

In [22]:
# 2. Create a boolean mask for customers with `roi >= 2.0`. 
roi_mask = roi >= 2.0
roi_mask

array([ True, False, False, False, False, False, False, False])

In [26]:
# 3. Use the mask to:  
#   - List their names  
#   - List their spend and revenue 
names_high_roi = names[roi_mask]
print(f"Names of High ROI: ", names_high_roi)
spend_high_roi = spend[roi_mask]
print(f"Spending with High ROI:", spend_high_roi)
revenue_high_roi = revenue[roi_mask]
print(f"Revenue of High ROI: ", revenue_high_roi)

Names of High ROI:  ['Ana']
Spending with High ROI: [200]
Revenue of High ROI:  [400]


In [27]:
# 4. Create a second mask for customers with `spend >= 200`.  
spending_mask = spend >= 200

In [28]:
# 5. Combine the two masks to find customers who have **roi \>= 2.0 AND spend \>= 200**. 
names_high_roi_and_spend = names[roi_mask & spending_mask]
print(f"Names of High ROI and Spend: ", names_high_roi_and_spend)

Names of High ROI and Spend:  ['Ana']


In [66]:
# Additional suggested solution, to do a combined mask
combined_mask = roi_mask & spending_mask
names_combined = names[combined_mask]
print(f"Names of customers with ROI >= 2.0 AND Spend >= 200: ", names_combined)

Names of customers with ROI >= 2.0 AND Spend >= 200:  ['Ana']


6. In a markdown cell, answer: *“What business insight do you get from this filtered group?”*

It filters to the names of people who manage to get a standard level of ROI. With this standard, we can identify and fund this person more to obtain more revenue, since the others who may not get the same conversion rate.

## 4\. Gradebook: Aggregations and broadcasting

In [29]:
# Reuse the “Gradebook” idea from class and extend it.

students = np.array(["S1", "S2", "S3", "S4", "S5"])
subjects = np.array(["Math", "Stats", "Python"])

scores = np.array([
    [75, 80, 85],
    [60, 65, 70],
    [90, 88, 92],
    [82, 79, 84],
    [70, 72, 78]
])

In [30]:
# 1. Compute each student’s **average score** (per row).  
average_scores = np.mean(scores, axis=1)  # axis=1 computes the mean across columns (subjects) for each row (student)
average_scores

array([80.        , 65.        , 90.        , 81.66666667, 73.33333333])

In [37]:
# 2. Compute each subject’s **average score** (per column).  
average_scores_subjects = np.mean(scores, axis=0)
average_scores_subjects

array([75.4, 76.8, 81.8])

In [38]:
# 3. Create `scores_centered` by subtracting the **subject mean** from each column (broadcasting).  
scores_centered = scores - average_scores_subjects  # This will broadcast the average scores of each subject across the respective columns
scores_centered

array([[ -0.4,   3.2,   3.2],
       [-15.4, -11.8, -11.8],
       [ 14.6,  11.2,  10.2],
       [  6.6,   2.2,   2.2],
       [ -5.4,  -4.8,  -3.8]])

In [36]:
# 4. For `scores_centered`, compute per-student averages again.  
scores_centered_avg = np.mean(scores_centered, axis=1)
scores_centered_avg

array([  2.        , -13.        ,  12.        ,   3.66666667,
        -4.66666667])

5. Compare which student looks best by **raw average** vs **centered average**.  

Raw Average: array([80. , 65. , 90. , 81.66666667, 73.33333333])

Centered Averages: array([2. , -13. ,  12. ,3.66666667, -4.66666667])

I feel that the question did not address the point well. I believe, with discussion with GPT, if we use RAW average, the student's high scores may pull up the average. However, if we take centered into consideration, we add the factor if a subject is easy or hard. With that, the new average may show a different picture.

## 5\. Reshape and flatten: From daily to weekly

In [39]:
# **Scenario:** You have 30 days of daily website visits, and you want to summarize them by week.

daily_visits = np.array([
    120, 130, 125, 140, 150,
    160, 170, 155, 145, 135,
    128, 132, 138, 142, 148,
    152, 158, 162, 168, 172,
    180, 190, 185, 175, 165,
    155, 145, 135, 125, 115
])

In [43]:
daily_visits.shape

(30,)

In [40]:
# 1. Reshape `daily_visits` into a `(6, 5)` array representing **6 weeks × 5 days**.  
weekly_visits = daily_visits.reshape(6, 5)  # Reshape the 1D array into a 2D array with 6 rows and 5 columns
weekly_visits


array([[120, 130, 125, 140, 150],
       [160, 170, 155, 145, 135],
       [128, 132, 138, 142, 148],
       [152, 158, 162, 168, 172],
       [180, 190, 185, 175, 165],
       [155, 145, 135, 125, 115]])

In [44]:
weekly_visits.shape

(6, 5)

In [45]:
# 2. Compute **total visits per week**.  
total_visits_per_week = np.sum(weekly_visits, axis=1)  # Sum across columns (days) for each row (week)
total_visits_per_week

array([665, 765, 688, 812, 895, 675])

In [46]:
# 3. Compute **average visits per day of the week** (e.g., all “day 1 of week” together, all “day 2 of week” together, etc.).  
average_visits_per_day = np.mean(weekly_visits, axis=0)  # Mean across rows (weeks) for each column (day)
average_visits_per_day

array([149.16666667, 154.16666667, 150.        , 149.16666667,
       147.5       ])

In [47]:
# 4. Flatten the reshaped array back to 1D and confirm it matches the original `daily_visits`.
weekly_visits_flattened = weekly_visits.flatten()  # Flatten the 2D array back to 1D
weekly_visits_flattened

array([120, 130, 125, 140, 150, 160, 170, 155, 145, 135, 128, 132, 138,
       142, 148, 152, 158, 162, 168, 172, 180, 190, 185, 175, 165, 155,
       145, 135, 125, 115])

In [48]:
weekly_visits_flattened.shape

(30,)

## 6\. Mini-project: Simple scoring model

In [50]:
# This mirrors the tiny matrix multiply example from class, but with more features.

# Each row: [page_views, time_on_site (minutes), past_purchases]
X = np.array([
    [10,  3.5,  0],
    [25,  5.0,  1],
    [40,  2.0,  0],
    [15, 10.0,  3],
    [30,  4.0,  2]
])

customers = np.array(["C1", "C2", "C3", "C4", "C5"])

In [51]:
# 1. Choose a weight vector `w = [w_views, w_time, w_purchases]` (for example `[0.1, 0.5, 1.0]`).  
w = np.array([0.1, 0.5, 1.0])

In [52]:
# 2. Compute a **score** for each customer using `scores = X @ w`.  
scores = X @ w  # This performs the matrix multiplication of X and w
scores

array([2.75, 6.  , 5.  , 9.5 , 7.  ])

In [54]:
# 3. Rank customers by score (highest first).
scores_ranked = np.argsort(scores)[::-1]  # Get the indices that would sort the scores in descending order. -1 reverses the order to get highest first.
scores_ranked

# what is returned is the rank of the customers, not the scores themselves. To get the scores in ranked order, we can do:
# scores_sorted = scores[scores_ranked]  # This will give us the scores sorted in descending order

array([3, 4, 1, 2, 0])

In [56]:
# 4. Change the weights to emphasize **past\_purchases** more than other features, and recompute. 
w_pastpurchases = np.array([0.1, 0.5, 2.0])  # Increased the weight for past_purchases

scores_pastpurchases = X @ w_pastpurchases
scores_pastpurchases

array([ 2.75,  7.  ,  5.  , 12.5 ,  9.  ])

In [57]:
scores_ranked_pp = np.argsort(scores_pastpurchases)[::-1]
scores_ranked_pp

array([3, 4, 1, 2, 0])

5. In a markdown cell, answer:  
   - Which customer is top-ranked before vs after changing weights?  
   - In what real-world situation might you prefer each weighting?

Same rank for both.

There is no change.
I believe if there is any case where the business wants to reach out to old customers who had been dormant, this may help?


## 7\. Stretch ideas (optional)

If you’re comfortable with the above, try:

- Generate 1,000 random test scores with `np.random.randn`, scale them to have mean 70 and standard deviation 10, then:  
  - Clip scores to the range `[0, 100]`  
  - Compute min, max, mean, and standard deviation  
- Simulate a small A/B test:  
  - Two groups of 20 users each  
  - Randomly generate conversions (0 or 1\) for each group  
  - Compute conversion rate per group using pure NumPy operations

In [58]:
# - Generate 1,000 random test scores with `np.random.randn`, scale them to have mean 70 and standard deviation 10, then:  

test_scores = np.random.randn(1000) * 10 + 70  # Scale to mean 70 and std dev 10. Why 10 and 70? Because the standard normal distribution has a mean of 0 and a standard deviation of 1, so multiplying by 10 scales the standard deviation to 10, and adding 70 shifts the mean to 70.
test_scores[:10]  # Show the first 10 test scores to confirm they look reasonable

array([83.06624099, 75.43145072, 73.94753196, 75.09199103, 84.92240399,
       77.94097701, 71.44602973, 79.61785742, 68.12249596, 67.04264258])

In [59]:
# In anticipation of the next step, to check if there are any values that are outside the expected range of 0 to 100, we can do:
# Check for any scores outside the range of 0 to 100
out_of_range_scores = test_scores[(test_scores < 0) | (test_scores > 100)]
print(f"Number of out-of-range scores: {len(out_of_range_scores)}")

Number of out-of-range scores: 0


In [60]:
#  - Clip scores to the range `[0, 100]` 
test_scores_clipped = np.clip(test_scores, 0, 100)  # This will set any values below 0 to 0 and any values above 100 to 100

In [61]:
#  - Compute min, max, mean, and standard deviation  
print(f"Min: {test_scores_clipped.min()}")
print(f"Max: {test_scores_clipped.max()}")
print(f"Mean: {test_scores_clipped.mean()}")
print(f"Standard Deviation: {test_scores_clipped.std()}")

Min: 36.3152183256249
Max: 96.27062170196206
Mean: 69.72017772187827
Standard Deviation: 10.098857748639583


In [62]:
"""
- Simulate a small A/B test:  
  - Two groups of 20 users each  
  - Randomly generate conversions (0 or 1\) for each group  
  - Compute conversion rate per group using pure NumPy operations
"""

# Simulate conversions for Group A and Group B
np.random.seed(42)  # Set a seed for reproducibility
group_A_conversions = np.random.choice([0, 1], size=20)  # Randomly generate 0s and 1s for Group A

# .choice() is a NumPy function that generates a random sample from a given 1-D array. In this case, we are using it to randomly select either 0 or 1 for each of the 20 users in Group A, simulating whether each user converted (1) or did not convert (0).
group_B_conversions = np.random.choice([0, 1], size=20)  # Randomly generate 0s and 1s for Group B

# Compute conversion rates
conversion_rate_A = group_A_conversions.mean()  # The mean of the binary array gives the conversion rate
conversion_rate_B = group_B_conversions.mean()  # The mean of the binary array gives the conversion rate
print(f"Conversion Rate for Group A: {conversion_rate_A:.2f}")
print(f"Conversion Rate for Group B: {conversion_rate_B:.2f}")


Conversion Rate for Group A: 0.35
Conversion Rate for Group B: 0.65


In [63]:
# Suggested solution:

"""
# Group sizes
n = 20

# Generate conversions (0 or 1)
# Using 0.3 probability for A and 0.4 for B
group_a = np.random.choice([0, 1], size=n, p=[0.7, 0.3])
group_b = np.random.choice([0, 1], size=n, p=[0.6, 0.4])

print(f"Group A Conv Rate: {group_a.mean():.2%}")
print(f"Group B Conv Rate: {group_b.mean():.2%}")
"""

# The reason why p=[0.7,0.3] is to give a probability of 0 and 1s appearing in the generated data. In this case, for Group A, there is a 70% chance of generating a 0 (no conversion) and a 30% chance of generating a 1 (conversion). For Group B, there is a 60% chance of generating a 0 and a 40% chance of generating a 1. This allows us to simulate different conversion rates for the two groups, which can be useful for testing the effectiveness of different strategies in an A/B test.